barretenberg/cpp/src/barretenberg/vm2/simulation/lib/hinting_dbs.cpp

```diff
@@ -0,0 +1,479 @@
+        .oldCheckpointId = old_checkpoint_id,
+        .newCheckpointId = get_checkpoint_id(),
+    };
+
+    // Update this db:
+    checkpoint_action_counter++;
+}
+
+uint32_t HintingContractsDB::get_checkpoint_id() const
+{
+    return checkpoint_stack.top();
+}
+
+void HintingContractsDB::dump_hints(ExecutionHints& hints)
+{
+    std::ranges::transform(contract_hints.contract_instances,
+                           std::back_inserter(hints.contractInstances),
+                           [](const auto& mapped_contract_instance) { return mapped_contract_instance.second; });
+
+    std::ranges::transform(contract_hints.contract_classes,
+                           std::back_inserter(hints.contractClasses),
+                           [](const auto& mapped_contract_class) { return mapped_contract_class.second; });
+
+    std::ranges::transform(contract_hints.bytecode_commitments,
+                           std::back_inserter(hints.bytecodeCommitments),
+                           [](const auto& mapped_bytecode_commitment) { return mapped_bytecode_commitment.second; });
+
+    std::ranges::transform(contract_hints.debug_function_names,
+                           std::back_inserter(hints.debugFunctionNames),
+                           [](const auto& mapped_debug_function_name) { return mapped_debug_function_name.second; });
+
+    std::ranges::transform(
+        create_checkpoint_hints,
+        std::back_inserter(hints.contractDBCreateCheckpointHints),
+        [](const auto& mapped_create_checkpoint_hint) { return mapped_create_checkpoint_hint.second; });
+
+    std::ranges::transform(
+        commit_checkpoint_hints,
+        std::back_inserter(hints.contractDBCommitCheckpointHints),
+        [](const auto& mapped_commit_checkpoint_hint) { return mapped_commit_checkpoint_hint.second; });
+
+    std::ranges::transform(
+        revert_checkpoint_hints,
+        std::back_inserter(hints.contractDBRevertCheckpointHints),
+        [](const auto& mapped_revert_checkpoint_hint) { return mapped_revert_checkpoint_hint.second; });
+}
+
+// Hinting MerkleDB starts.
+const AppendOnlyTreeSnapshot& HintingRawDB::get_tree_info(world_state::MerkleTreeId tree_id) const
+{
+    auto roots = db.get_tree_roots();
+    return get_tree_info_helper(tree_id, roots);
+}
+
+SiblingPath HintingRawDB::get_sibling_path(world_state::MerkleTreeId tree_id, index_t leaf_index) const
+{
+    auto tree_info = get_tree_info(tree_id);
+    auto path = db.get_sibling_path(tree_id, leaf_index);
+    GetSiblingPathKey key = { tree_info, tree_id, leaf_index };
+    merkle_hints.get_sibling_path_hints[key] =
+        GetSiblingPathHint{ .hintKey = tree_info, .treeId = tree_id, .index = leaf_index, .path = path };
+
+    return path;
+}
+
+GetLowIndexedLeafResponse HintingRawDB::get_low_indexed_leaf(world_state::MerkleTreeId tree_id, const FF& value) const
+{
+    auto tree_info = get_tree_info(tree_id);
+    auto resp = db.get_low_indexed_leaf(tree_id, value);
```

Comment:
If possible I'd use less `auto` if the type is not clear from context. I.e., consider explicitly writing the type